src/officehours_api/consumers.py

```diff
@@ -0,0 +1,312 @@
+
+from officehours_api.models import Queue, Meeting, Profile
+from officehours_api.permissions import is_host
+from officehours_api.serializers import (
+    QueueHostSerializer, QueueAttendeeSerializer, UserSerializer
+)
+
+
+class QueueConsumer(JsonWebsocketConsumer):
+    _queue_id: int
+    _user: User
+
+    @staticmethod
+    def get_group_name(queue_id):
+        return f'queue_{queue_id}'
+
+    @property
+    def queue_id(self):
+        return self._queue_id
+
+    @property
+    def group_name(self):
+        return self.get_group_name(self.queue_id)
+
+    @property
+    def user(self):
+        return self._user
+
+    def connect(self):
+        self._queue_id = int(self.scope['url_route']['kwargs']['queue_id'])
+        self._user = self.scope["user"]
+        try:
+            queue = Queue.objects.get(pk=self.queue_id)
+        except Queue.DoesNotExist:
+            self.accept()
+            self.close(code=4404)
+            return
+
+        QueueSerializer = (
+            QueueHostSerializer
+            if is_host(self.user, queue)
+            else QueueAttendeeSerializer
+        )
+
+        async_to_sync(self.channel_layer.group_add)(
+            self.group_name,
+            self.channel_name
+        )
+        self.accept()
+        queue_data = QueueSerializer(queue, context={'user': self.user}).data
+        self.send_json({
+            'type': 'init',
+            'content': queue_data,
+        })
+
+    def disconnect(self, close_code):
+        async_to_sync(self.channel_layer.group_discard)(
+            self.group_name,
+            self.channel_name
+        )
+
+    def queue_update(self, event):
+        queue = Queue.objects.get(pk=self.queue_id)
+        QueueSerializer = (
+            QueueHostSerializer
+            if is_host(self.user, queue)
+            else QueueAttendeeSerializer
+        )
+        queue_data = QueueSerializer(
+            queue,
+            context={'user': self.user},
+        ).data
+        self.send_json({
+            'type': 'update',
+            'content': queue_data,
+        })
+
+    def queue_deleted(self, event):
+        self.send_json({
+            'type': 'deleted',
+        })
+
+
+def send_queue_update(queue_id: int, channel_layer=None):
+    channel_layer = channel_layer or get_channel_layer()
+    async_to_sync(channel_layer.group_send)(
+        QueueConsumer.get_group_name(queue_id),
+        {
+            'type': 'queue.update',
+        }
+    )
+
+
+def send_queue_delete(queue_id: int, channel_layer=None):
+    channel_layer = channel_layer or get_channel_layer()
+    async_to_sync(channel_layer.group_send)(
+        QueueConsumer.get_group_name(queue_id),
+        {
+            'type': 'queue.deleted',
+        }
+    )
+
+
+@receiver(post_save, sender=Queue)
+def trigger_queue_update(sender, instance: Queue, created, **kwargs):
+    if instance.deleted:
+        return
+    transaction.on_commit(lambda: send_queue_update(instance.id))
```

Comment:
Changes driven from the admin panel use `transaction.atomic()`, which causes a race condition where the message event handlers (e.g. `queue_update`) may outrun the transaction commit, causing out-of-date data to be sent. `on_commit` makes sure the messages aren't triggered until after the commit.